beacon-chain/sync/pending_blocks_queue.go

```diff
@@ -270,28 +289,68 @@ func (s *Service) deleteBlockFromPendingQueue(slot uint64, b *ethpb.SignedBeacon
 		newBlks = append(newBlks, blk)
 	}
 	if len(newBlks) == 0 {
-		delete(s.slotToPendingBlocks, slot)
-		return
+		s.slotToPendingBlocks.Delete(slotToCacheKey(slot))
+		return nil
+	}
+
+	if err := s.slotToPendingBlocks.Replace(slotToCacheKey(slot), newBlks, gcache.DefaultExpiration); err != nil {
+		return err
 	}
-	s.slotToPendingBlocks[slot] = newBlks
 	delete(s.seenPendingBlocks, r)
+	return nil
 }
 
 // Insert block to the list in the pending queue using the slot as key.
 // Note: this helper is not thread safe.
-func (s *Service) insertBlockToPendingQueue(slot uint64, b *ethpb.SignedBeaconBlock, r [32]byte) {
+func (s *Service) insertBlockToPendingQueue(slot uint64, b *ethpb.SignedBeaconBlock, r [32]byte) error {
 	mutexasserts.AssertRWMutexLocked(&s.pendingQueueLock)
 
 	if s.seenPendingBlocks[r] {
-		return
+		return nil
 	}
 
-	_, ok := s.slotToPendingBlocks[slot]
-	if ok {
-		blks := s.slotToPendingBlocks[slot]
-		s.slotToPendingBlocks[slot] = append(blks, b)
-	} else {
-		s.slotToPendingBlocks[slot] = []*ethpb.SignedBeaconBlock{b}
+	if err := s.addPendingBlockToCache(b); err != nil {
+		return err
 	}
+
 	s.seenPendingBlocks[r] = true
+	return nil
+}
+
+// This returns signed beacon blocks given input key from slotToPendingBlocks.
+func (s *Service) pendingBlocksInCache(slot uint64) []*ethpb.SignedBeaconBlock {
+	k := slotToCacheKey(slot)
+	value, ok := s.slotToPendingBlocks.Get(k)
+	if !ok {
+		return []*ethpb.SignedBeaconBlock{}
+	}
+	blks, ok := value.([]*ethpb.SignedBeaconBlock)
+	if !ok {
+		return []*ethpb.SignedBeaconBlock{}
+	}
+	return blks
+}
+
+// This adds input signed beacon block to slotToPendingBlocks cache.
+func (s *Service) addPendingBlockToCache(b *ethpb.SignedBeaconBlock) error {
+	if b == nil || b.Block == nil {
+		return errors.New("nil block")
+	}
+
+	blks := s.pendingBlocksInCache(b.Block.Slot)
+	blks = append(blks, b)
+	k := slotToCacheKey(b.Block.Slot)
+	return s.slotToPendingBlocks.Add(k, blks, gcache.DefaultExpiration)
```

Comment:
The value of  `gcache.DefaultExpiration` is 0, and since the method then uses a previously initialized default expiration of `1 * time.Second` , no pending block lives in the cache long enough to be useful.

Reply:
Also another issue, if a key already exists for the particular slot in the cache, this method will instead return an error. 
https://github.com/patrickmn/go-cache/blob/master/cache.go#L90

Reply:
Also another side-effect is that expiration times are reset upon each addition to the cache. Which lends itself to being a DOS vector.

Reply:
Updates:
* Replaced `Add` with `Set`
* I'm inclined to say this side-effect is ok since we are only allowing 3 blocks per slot, here I also can reduce the expiration time in proportion to block count if you think it's necessary